### PR TITLE
Sort the structs returned by /api/topology

### LINKED ISSUE #561: Containers > By Image / By Hostname swaps around
We probably need to return these in the backend in an array rather than a map

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #561 

Also dedupe some of of the structs in the app, topologyView -> APITopolgyDesc, optionValue -> APITopologyOption.

### CODE REVIEW

File: app/api_topologies.go [hunk 1]
```diff
@@ -31,6 +139,52 @@ type topologyStats struct {
...
+
+func (r *registry) add(ts ...APITopologyDesc) {
+	r.Lock()
+	defer r.Unlock()
+	for _, t := range ts {
+		t.URL = apiTopologyURL + t.id
+
+		if t.parent != "" {
+			parent := r.items[t.parent]
+			parent.SubTopologies = append(r.items[t.parent].SubTopologies, t)
```
Comment (paulbellamy) [inline app/api_topologies.go]: Probably `append(parent.SubTopologies, t)`, no?

File: app/api_topologies.go [hunk 2]
```diff
@@ -31,6 +139,52 @@ type topologyStats struct {
...
+			sort.Sort(byName(parent.SubTopologies))
+			r.items[t.parent] = parent
+		}
+
+		r.items[t.id] = t
```
Comment (paulbellamy) [inline app/api_topologies.go]: Shouldn't we only do this if it's not a sub-toplogy?

File: app/api_topologies.go [hunk 3]
```diff
@@ -31,6 +139,52 @@ type topologyStats struct {
...
+	for _, desc := range r.items {
+		if desc.parent != "" {
+			continue
+		}
+		descs = append(descs, desc)
+	}
+	sort.Sort(byName(descs))
+	for _, desc := range descs {
+		f(desc)
+	}
```
Comment (paulbellamy) [inline app/api_topologies.go]: Why not `sort.Sort(byName(t.items))` in `add`?

File: app/api_topologies.go [hunk 4]
```diff
@@ -31,6 +139,52 @@ type topologyStats struct {
 	FilteredNodes      int `json:"filtered_nodes"`
 }
 
+type byName []APITopologyDesc
```
Comment (paulbellamy) [inline app/api_topologies.go]: Let's move this up next to APITopologyDesc, to group it with the thing it sorts.

File: app/api_topologies.go [hunk 5]
```diff
@@ -31,6 +139,52 @@ type topologyStats struct {
...
+		descs = append(descs, desc)
+	}
+	sort.Sort(byName(descs))
+	for _, desc := range descs {
+		f(desc)
+	}
+}
+
 // makeTopologyList returns a handler that yields an APITopologyList.
 func makeTopologyList(rep xfer.Reporter) func(w http.ResponseWriter, r *http.Request) {
```
Comment (paulbellamy) [inline app/api_topologies.go]: This would be cleaner as a method on `registry`.

File: app/api_topologies.go (hunk 1 above)
Comment (tomwilkie) [inline app/api_topologies.go]: Ah yes, good catch.

File: app/api_topologies.go (hunk 2 above)
Comment (tomwilkie) [inline app/api_topologies.go]: Unfortunately not; sub topologies are still looked up by id (they are not nested in their URL).

File: app/api_topologies.go (hunk 3 above)
Comment (tomwilkie) [inline app/api_topologies.go]: Add puts them in a map, not a list.

I considered having add put them in a list as well, and keeping that sorted, but it wasn't clear to me if the k8s topologies were added multiple times, and if add was supposed to dedupe.

File: app/api_topologies.go (hunk 4 above)
Comment (tomwilkie) [inline app/api_topologies.go]: Can do.  I put it there to be near the functions, and typically we don't like to mix functions and type definitions.

File: app/api_topologies.go (hunk 5 above)
Comment (tomwilkie) [inline app/api_topologies.go]: Will do.

File: app/api_topologies.go (hunk 2 above)
Comment (paulbellamy) [inline app/api_topologies.go]: Ah, yeah.

File: app/api_topologies.go (hunk 4 above)
Comment (paulbellamy) [inline app/api_topologies.go]: I meant move the whole `byName` and it's fns. Either way, though.

File: app/api_topologies.go (hunk 3 above)
Comment (tomwilkie) [inline app/api_topologies.go]: Discussed offline, not worth it.

**Comment by tomwilkie** [comment]
@paulbellamy PTAL

**Comment by paulbellamy** [comment]
LGTM

**Pull request merged**